_posts/2016-07-11-value-of-automated-e2e-testing.md

```diff
@@ -0,0 +1,58 @@
+
+![A bunch of icons](/img/value-of-e2e-testing/ikonit.JPG)
+
+##The First Try with Wicket Unit Tests
+
+The first automated e2e-tests were written using [Wicket Unit Tests](https://cwiki.apache.org/confluence/display/WICKET/Unit+Test). The unit tests utilized WicketTester, which runs the application without the need to actually run a web server. You can validate the web page opening, click some buttons, fill fields, run the business logic and validate the web page after. The Wicket tests are pretty quick to run and fairly simple to write. But there are also disadvantages, the tests are not actually testing the full pipe, because the lack of web server and browser. Also, the Wicket page tester does not create a picture of the page, meaning that in a case of exception or assertion error, the actual web page cannot be seen. It is like trying to find a needle in a dark room. Furthermore, the actual page structure can be seen only in debug output, i.e. finding the components to assert is not always easy. 
+
+However, despite the impediments, these tests were found very helpful. The object was reached, the tests checked that the icons were visible and clickable in the specified situations, and even more, if you broke functionality by accident it was exposed by a test.
+
+##The New Era with JBehave and Selenium
+
+After a while, we got a testing specialist in our project, and a new era began. He started to build real automated e2e-tests with JBehave and Selenium. JBehave enables writing the tests as stories in text files. Each story file has a matching Java class implementing the steps described in the story. The steps and the actual test code are mapped together with annotations. The web page actions, button clicks, element searches and so on are written in separate bridge classes (e.g. MainItemPageBridge.java). The tests are run with the help of Selenium Web Driver, which links the actual web page and the test code together.
+
+![An example of JBehave and Selenium in action](/img/value-of-e2e-testing/jbehave.png)
+
+At first, there was a huge gap between code implemented and code tested with JBehave tests. There were a couple of guys just writing tests for the already implemented features. However, soon it was clear, that the gap would never be surpassed this way. Everyone in the project should learn writing the new type of tests. In my opinion, in the beginning it was slow and not very easy but bit by bit it became a natural way to write tests. Of course, the test framework developed side by side with the tests, because we wrote a lot of common utility methods and tools.
```

Comment:
guys? All male testing panel? :) Tuo on varmasti ihan totta, mutta voisi löytyä joku muu ilmaus?
bit by bit -> step by step it?
...the test framework was developed (?)...